### PR TITLE
[`v6`] fix: upcast cross-encoder logits to float32 before activation (HPS)

### LINKED ISSUE #3894: Similarity and cross-encoder scores computed below float32 create spurious ties in low-precision inference
### Summary

When a model runs in **float16/bfloat16** (e.g. `model_kwargs={"torch_dtype": torch.float16}`), sentence-transformers computes the final relevance score in that same low precision in two paths. Reduced-mantissa formats coarsely bucket values in the narrow score range, so **distinct query–document scores collapse into identical values ("spurious ties")**. Ranking then falls back to arbitrary tie-breaking (index order), which inflates the variance and bias of retrieval metrics and can even flip model rankings.

This is analyzed in [*Reliable Evaluation Protocol for Low-Precision Retrieval*](https://aclanthology.org/2026.acl-short.33) (Yang et al., ACL 2026). The remedy, **High-Precision Scoring (HPS), upcasts only the final scoring op to float32**, leaving the forward pass in low precision. It is value-preserving (every fp16/bf16 value is exactly representable in fp32), zero-parameter, and adds negligible cost.

MTEB recently merged the same two fixes: embeddings-benchmark/mteb#4933 (similarity primitives) and embeddings-benchmark/mteb#4934 (cross-encoder logits), from issue embeddings-benchmark/mteb#4932.

### Where it happens on current `main`

1. **Similarity primitives** — `sentence_transformers/util/tensor.py::_convert_to_tensor` already upcasts **sparse** tensors to float32, but leaves dense tensors at their incoming dtype. A fp16/bf16 model with `encode(..., convert_to_tensor=True)` returns low-precision embeddings, so `model.similarity()` / `util.cos_sim` / `dot_score` (and every evaluator built on them, e.g. `InformationRetrievalEvaluator`) scores in low precision. Note that `encode(..., convert_to_numpy=True)` upcasts **bfloat16** on conversion (numpy has no bf16) but keeps **float16** as fp16 numpy arrays, which are converted back to fp16 tensors when scoring.

2. **Cross-encoder scoring** — `CrossEncoder.predict` applies `activation_fn` (sigmoid) / softmax directly to the model's logits, i.e. in model dtype. The `convert_to_numpy` tail does call `.float()`, but only *after* the activation has already bucketized the (0, 1) probability range.

### Illustration

```python
import torch
q, d = torch.randn(1, 32), torch.randn(200, 32)
qn = torch.nn.functional.normalize
naive = qn(q.bfloat16(), dim=1) @ qn(d.bfloat16(), dim=1).T  # score in bf16
hps   = qn(q.bfloat16().float(), dim=1) @ qn(d.bfloat16().float(), dim=1).T  # HPS
print(torch.unique(naive).numel(), torch.unique(hps).numel())  # e.g. 184 vs 200

logits = torch.randn(200).bfloat16()  # cross-encoder logits
p_bf16 = torch.sigmoid(logits)
p_hps  = torch.sigmoid(logits.float())
print(torch.unique(p_bf16).numel(), torch.unique(p_hps).numel())  # e.g. ~150 vs 200
```

Fewer unique scores = spurious ties, broken arbitrarily during ranking.

### Proposed fix

Apply HPS at both sites, exactly as merged in MTEB:

- Upcast sub-float32 dense floating tensors to float32 in `_convert_to_tensor` — this mirrors the float32 upcast the same function already applies to sparse inputs.
- Upcast cross-encoder logits to float32 before the activation in `CrossEncoder.predict`.

Both are no-ops for float32 inputs, so full-precision results are unchanged. Under AMP/autocast training the affected ops are re-cast per autocast policy, so training behavior is effectively unchanged as well.

I've prepared two small, separate PRs and am opening them alongside this issue. Happy to adjust scope if maintainers prefer.

Disclosure: I'm the first author of the referenced paper; the same fixes were merged into MTEB this week (embeddings-benchmark/mteb#4933, embeddings-benchmark/mteb#4934).

### PR DESCRIPTION
**Pull request by KisuYang**
Fixes #3894 (cross-encoder part).

CrossEncoder.predict applies the activation function (sigmoid/softmax) directly to the logits produced by the model. When the model runs in float16/bfloat16, the activation is computed in that same low precision; the reduced mantissa coarsely buckets the (0, 1) probability range, so distinct query-document relevance scores collapse into spurious ties. Reranking then depends on arbitrary tie-breaking, which inflates the variance and bias of downstream retrieval metrics. The existing convert_to_numpy path already calls .float(), but only after the activation has bucketized the scores.

Upcast sub-float32 logits to float32 before the activation (High-Precision Scoring). The upcast is value-preserving (every float16/bfloat16 value is exactly representable in float32) and a no-op for float32 logits, so full-precision results are unchanged.

Reference
- embeddings-benchmark/mteb#4934
- [*Reliable Evaluation Protocol for Low-Precision Retrieval*](https://aclanthology.org/2026.acl-short.33) (Yang et al., ACL 2026)

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hello!

Thanks for this PR as well! Same story as #3892: the diagnosis is right, and the fix sits in exactly the right place, before the activation. Once sigmoid has run in bfloat16, the saturation damage is unrecoverable: the later `.float()` that the numpy path already did just re-encodes the already-bucketed values in a wider container.

To quantify it, I evaluated `cross-encoder/ettin-reranker-32m-v1` in bfloat16 on NanoBEIR (msmarco, nq, fiqa2018), reranking 100 candidates per query, with and without this fix:

- With the model's default activation, which for this particular model is `Identity` (raw logits), the metrics are identical with and without the fix: bfloat16 logits keep roughly 0.4 percent relative spacing, which is enough resolution for 100 candidates.
- With `activation_fn=torch.nn.Sigmoid()`, i.e. what a default single-label reranker uses, the difference is dramatic:

| | with fix | without fix |
|---|---|---|
| NanoBEIR mean NDCG@10 | **0.6795** | 0.1849 |
| NanoBEIR mean MRR@10 | **0.6797** | 0.3986 |
| unique scores over 15,040 pairs | 710 | 270 |

The mechanism is sigmoid saturation: a confident reranker pushes many of each query's candidates toward 1.0, where bfloat16 has a spacing of about 0.0039, so whole blocks of top candidates collapse onto the same value and the positive document lands at an arbitrary position inside the tie block. NanoMSMARCO NDCG@10 falls from 0.7093 to 0.0965, essentially randomized. Your test pins exactly this failure mode.

One small addition on my side: the new test casts a model to bfloat16 and runs CPU inference, which can hard-crash part of our Windows CI fleet on torch 2.13 and newer, so I added our `skip_bfloat16_cpu_crash` marker next to your older-torch xfail, one guard per end of the torch range.

Worth noting for completeness: even with the fix, the 710 unique scores show the remaining ceiling. A bfloat16 model produces logits on the bfloat16 grid, so the fix removes the damage the library adds on top, not the rounding the model itself already did.

Thanks again, this and #3892 make low-precision inference behave the way everyone assumed it already did!

- Tom Aarsen

**Comment by KisuYang** [comment]
Thank you @tomaarsen for the thorough verification. The NanoBEIR numbers demonstrate the impact better than I could have, and the reshape in #3892 makes perfect sense. Happy to see these land in v6.0.
